rgw: new vault transit logic for SSE-C

mdw-at-linuxbox opened the pull request:
This is a set of commits for the following:

rgw/vault/transit
  Use transit the way it's supposed to be: have it coin a "datakey" instead of just exporting the key and (in effect) treating it exactly as if it were a kv key store.
rgw/kmip/kms
 Support a kmip server as a kms.

This includes and supersedes earlier fixes I had previously presented in PR#33996 .  I will be closing those in favour of this PR.

tchaikov commented on .gitmodules:
```diff
@@ -77,3 +77,7 @@
+[submodule "src/libkmip"]
+	path = src/libkmip
+	url = https://github.com/ceph/libkmip
+	branch = wip-morestuff
```
i'd suggest use a branch with a more meaningful name so

- it is less likely to be removed be accident
- we can switch to another branch when we need to bump up the submodule in case we the new branch is not a fast-forward of the old branch.

mdw-at-linuxbox commented:
I pushed an update that /1/ fixes the libkmip branch name, /2/ fixes another problem that stopped teuthology from testing using pykmip, and /3/ is rebased on a newer master  So far, the newer code builds on crimson, but has some new problem that broke 'make check' I think differently than before.  I've got a teuthology run queued up - I'll fix the make check problem and make an attempt to fix whatever ails the teuthology part once that run completes.

mdw-at-linuxbox commented:
I'm currently trying to get this PR to pass its various checks.  I've just pushed a fix for documentation, "make check", and teuthology.  Also rebased on a slightly newer copy of master.

mdw-at-linuxbox commented:
The current "make check" problems appear to have nothing to do with this PR: one is with dpkg-preconfigure failing - system problem? The other is with "unittest_rgw_amqp" - might be a system problem too.  My copy of the same test completes with no issues.

mdw-at-linuxbox commented:
I pushed a new version.  Mostly teuthology fixes.  Fix for barbican teuthology y21 problem, keystone centos-i bindep fix.  Vault unzip on centos-8.  pykmip: make keys w/ python2 on ubuntu.  "make check" / kms: fix unit test so that it compiles and works.  Also rebased on latest master.

mdw-at-linuxbox commented:
Various sources have indicated an interest in breaking this PR into multiple parts.  The cost of that is a possibility of taking an extra week per PR to get it committed -- potentially a month more of elapsed time to match this PR.

This PR segments most naturally into the following:
kmip (14 commits)
vault transit (13 commits)
vault unzip (1) / cenotos-8
keystone bindep (1) / centos-8
barbican y2021 (1)

kmip & vault fixes are many commits because I tried to structure them into smaller easier to understand commits.  vault should be clean in this respect; kmip may have more bug fixes in this respect.
Vault transit is likely at least textually dependent on kmip.  They affect the same code, and I did not want to even think about factoring out anything in common with kmip to work on apply on top of previously commited vault transit fix.

vautl unziip, keystone bindep, and barican y2021 are each one commit, and can be done in any order.  vault transit/kmip aren't dependent on these -- except that teuthology runs against kmip/vault transit will have known failures (and will hide any regression failures) without these fixes.

vault unzip, keystone bindep fix centos-8 issues; not having these breaks my testing.  Most upstream testing uses ubuntu, and the unzip cases could be masked by non-standard system configuration, so these could appear as only sporadic failures upstream.

mattbenjamin commented on src/rgw/rgw_kmip_client.cc:
```diff
@@ -0,0 +1,82 @@
+	int i;
+	if (out)
+		free(out);
+	out = 0;
```
nullptr, these days

mdw-at-linuxbox replied:
nit: will fix

mattbenjamin commented on src/rgw/rgw_kmip_client.cc:
```diff
@@ -0,0 +1,82 @@
+		for (i = 0; i < outlist->string_count; ++i) {
+			free(outlist->strings[i]);
+		}
+		free(outlist->strings);
```
why is this on the heap? as opposed to, vector<string> or something?

mdw-at-linuxbox replied:
Ultimately it's a list of lists, so it's going to come off the heap no matter what.  This could all be coded using std::vector<std::string>;   That's probably not too costly, but there are other elements of RGWKMIPTransceiver where it gets more complicated.

So in src/rgw/rgw_kmip_client.h there's a declaration for RGWKMIPTransceiver.
outkey could be a string but that hides something I wish I could instead make more visible to the code: this isn't a utf8 string, it might contain embedded nulls, and it ought to be zeroized on free.  That's all true until it hits KmipGetTheKey::get_key_for_uniqueid where it gets turned into a std::string, at which point it's up tto the compiler if it actually does the "actual_key.replace" zeroize in src/rgw/rgw_crypt.cc.
out, name, unique_id are "char *"; there's logic that depends on being able to tell the difference between an unset value and a 0 length value which would need adjusting.

So doable, but complications.

mattbenjamin replied:
It's true, std::vector<std::string> uses the free store just as aggressively;  It does have the advantage of raii.  I'm not advocating you change this, honestly.  Just what was going through my mind reading the code.

mattbenjamin commented on src/rgw/rgw_kms.cc:
```diff
@@ -383,6 +492,28 @@ static int get_actual_key_from_vault(CephContext *cct,
+                                     std::string& actual_key)
+{
+  std::string secret_engine = RGW_SSE_KMS_KMIP_SE_KV;
+#if 0
```
what's going on in the #if 0 block?

mdw-at-linuxbox replied:
cruft; will remove.

I wasn't sure if I needed this as debugging - for now I definitely don't.

mattbenjamin commented on qa/tasks/pykmip.py:
```diff
@@ -0,0 +1,520 @@
+                    body=key1_json
+                )
+            except:
+                log.info("catched exception!")
```
catched?

mdw-at-linuxbox replied:
The same phrase is in the barbican task, which I cloned in the best tradition of John Frum.

But you'll be pleased that this disappears in a later kmip commit in this PR, which finishes implementing create_secrets.

mattbenjamin commented on src/rgw/rgw_crypt.cc:
```diff
@@ -25,6 +30,273 @@
+    rapidjson::StringStream isw(iline.c_str());
+    if (!iline.length())
+        d.SetObject();
+//    else if (qflag)
```
what's happening with the commented code?

mdw-at-linuxbox replied:
This is meant to be a placeholder to mark one way where one solution for integer handling could be enabled. Integers have real problems in canonical json, which is why it's disabled for now.  If we thought this way was right, it could be a future config option.

mattbenjamin commented on src/rgw/rgw_crypt.cc:
```diff
@@ -25,6 +30,273 @@
+	s->err.message = "context: internal error adding defaults";
+        return -ERR_INVALID_REQUEST;
+    }
+    b = make_everything_canonical(d, allocator, ccs, options) == mec_error::success;
```
make_everything_canonical() is ... canonicalize()?

mdw-at-linuxbox replied:
This one crawls down through a tree structure making everything canonical; hence that word "everything".  The rest of this code uses "canonical" in various aspects - I don't see much win to inconsistently verbalizing it here.

mattbenjamin commented on src/rgw/rgw_kms.cc:
```diff
@@ -333,14 +345,14 @@ class KmipSecretEngine: public SecretEngine {
-  int decode_secret(JSONObj* json_obj, std::string& actual_key){
-    return -EINVAL;
-  }
+//  int send_request(std::string_view key_id, JSONParser* parser) override
```
same here, what's happening with commented code?

mdw-at-linuxbox replied:
nit; will remove.

This was "old" code that I left in case I had to put it back in.

mattbenjamin commented on src/rgw/rgw_kms.cc:
```diff
@@ -13,13 +13,91 @@
+ * carefully zeros out all memory before returning it to
+ * the system.
+ */
+#define ALIGNTYPE double
```
there are some standards-track pool allocators, but...

mdw-at-linuxbox replied:
The framework is a requirement of rapidjson; any other implementation would require the same code overhead.  The actual allocator here can take advantage of how it's being used to avoid tracking the lengths of all the individual pieces, which is a win because many of these pieces will be very small.  Also note: no need for multiple-thread support.  General purpose allocators can't know that, so they have to have more code & overhead.  Most of the allocators I've found handle problems I don't care about, but fail to give me the hooks I want.  I suspect the glue code for such an allocator into this framework would be much larger than actual allocator here.

mattbenjamin commented:
@mdw-at-linuxbox are there teuthology runs?

mattbenjamin reviewed (commented):
lgtm, would like review from original authors, if possible

mdw-at-linuxbox commented on src/common/options.cc:
```diff
@@ -7049,6 +7049,38 @@ std::vector<Option> get_rgw_options() {
+    .set_default("")
+    .set_description("connect using client certificate"),
+
+    Option("rgw_crypt_kmip_kms_key_template", Option::TYPE_STR, Option::LEVEL_ADVANCED)
```
this could certaily use a description.  I should also find out where description and long_description are used - I've never managed to convince myself this is actually useful.

But, *cough*, speaking of documentation.  Yes, there should be documentation on kmip, somehow it didn't happen.  Yes indeed, sigh.

mdw-at-linuxbox commented:
I've been having trouble getting this all to pass teuthology.

Here's my latest run,
http://pulpito.front.sepia.ceph.com/mwatts-2021-01-27_21:05:08-rgw:crypt-mdw-transit-14-distro-basic-smithi/
I looked at this just now; this is the first one where everything actually passed. Yay.

This is just rgw/crypt, not all the rest.  This is run against exactly the version of ceph from this PR, but with a qa suite from a slightly more advanced copy.

todo:
1/ address at least the nits above
2/ rebase to latest master
3/ push to ci, build, q full run against teuthology.

mdw-at-linuxbox commented:
I've pushed an update: addresses minor nits above, includes new documentation on kmip configuration, and teuthology fixes.

I started two teuthology runs against this; first one (rgw/crypt) is complete already (no errors):
http://pulpito.front.sepia.ceph.com:80/mwatts-2021-02-04_17:12:18-rgw:crypt-mdw-transit-21-distro-basic-smithi/
http://pulpito.front.sepia.ceph.com:80/mwatts-2021-02-04_18:29:33-rgw-mdw-transit-21-distro-basic-smithi/

mdw-at-linuxbox commented:
I've pushed a new version of this.  It is rebased on the very latest master, and I reordered the kmip doc bit to apply before any of the vault changes.  There are no other changes from what I previously pushed and have running in teuthology.  The first part exactly matches what I just pushed to the kmip PR (#33996).  Once the kmip bits are committed, I intend to factor out the kmip bits from the vault changes here, which I believe have at least textual dependencies on the kmip work.

I ran the previous version through teuthology - 2 runs: first against rgw/crypt - which completed successfully as I reported above.  The second is against "rgw".  As of "now"; one test of 145 completed, all the others are still queued.

cbodley commented on src/rgw/rgw_kms.cc:
```diff
@@ -189,67 +316,405 @@ class TransitSecretEngine: public VaultSecretEngine {
+      lderr(cct) << "ERROR: vault transit secrets engine : parameter "
+	<< e.first << "=" << e.second << " ignored" << dendl;
+    }
+    if (compat == -1) {
+      std::string_view v { cct->_conf->rgw_crypt_vault_prefix };
+      if (string_ends_maybe_slash(v,"/export/encryption-key")) {
+	compat = 2;
+      } else {
+	compat = 0;
+      }
```
what do these values of `compat` mean? consider giving them names with enum

cbodley commented on doc/radosgw/vault.rst:
```diff
@@ -255,12 +393,44 @@ restricted to fetch KV keys as follows::
-``http://vault-server:8200/v1/transit/export/encryption-key``.
+``https://vault-server:8200/v1/transit``.
+
+Transit engine compatibility support
+------------------------------------
+The transit engine has compatibility support for previous
+versions of ceph, which used the transit engine as a simple key store.
+
+There is a a "compat" option which can be given to the transit
+engine to configure the compatibility support,
```
can you please add a note about this to `PendingReleaseNotes`? we want existing users to know about the new/recommended configuration

cbodley reviewed (commented):
i like the way this handles backward compatibility, and it's communicated well it in the docs 👍

mdw-at-linuxbox commented:
I pushed a new version - rebased and addresses one set of nits for kmip.  I've got a build going throuch ceph-ci, but I won't bother running it through teuthology since there appear to be more nits to wack.  The run I scheduled in teuthology for the previous version completed - only 3 failures, all I believe unrelated to these changes.  One with "sts", and 2 involving something about pg availability.  I hope to have another update soon.

mdw-at-linuxbox commented:
I've just pushed an update that I believe addresses most of the nits people raised here and w/ kmip.  I also rebased it on the latest master.  A local build succeeded; I have it q'd up at ceph-ci, will try running it through teuthology.

mdw-at-linuxbox commented:
Apparently github is picky about merging.  I've rebased and pushed yet another version.  No interesting changes.

The previous version made its way through teuthology.  Runs here,
http://pulpito.front.sepia.ceph.com:80/mwatts-2021-03-03_00:18:06-rgw:crypt-mdw-transit-28-distro-basic-smithi/

http://pulpito.front.sepia.ceph.com:80/mwatts-2021-03-03_01:47:23-rgw-mdw-transit-28-distro-basic-smithi/
rgw/crypt completed without error.  I haven't looked at the longer rgw run in detail; I hope the few failures I see there are all "expected".

cbodley commented:
build failure from make check:
```
/home/jenkins-build/build/workspace/ceph-pull-requests/src/rgw/rgw_crypt.cc: In function 'const string& get_tenant_or_id(req_state*)':
/home/jenkins-build/build/workspace/ceph-pull-requests/src/rgw/rgw_crypt.cc:216:48: error: binding reference of type 'std::string&' {aka 'std::__cxx11::basic_string<char>&'} to 'const string' {aka 'const std::__cxx11::basic_string<char>'} discards qualifiers
  216 |     std::string &tenant{ s->user->get_tenant() };
      |                                                ^
```

mdw-at-linuxbox commented:
I pushed a new version that factors out the kmip support.  I don't know yet if that addresses "make check" above.

mdw-at-linuxbox commented:
I pushed yet another version.  This one fixes a problem with "const"ness that was breaking "make check".
